### ISSUE TITLE
Production dialog displayed twice

### ISSUE THREAD

**Issue by ssoloff** [triplea-game/triplea#2184]
(Originally reported by @prastle on the [forum](https://forums.triplea-game.org/post/2461).)

As of at least TripleA 1.9.0.0.6028, the Production dialog is displayed twice in sequence.  The first time it is displayed, it has no content.  After closing the first dialog, a second Production dialog is opened with the expected content.

This issue appears to have been introduced in c6fa60d.  It's not clear, but it may have been due to a merge.  It's obvious from that current state of the `ProductionPanel` code that the Production dialog will be displayed twice, as reported above.  Reverting the `ProductionPanel` changes from c6fa60d appears to fix this issue.

However, there's a latent NPE hiding here.  The code prior to c6fa60d checked if the parent frame was non-`null` before creating the dialog.  In the case where the parent frame was `null`, an NPE would have been thrown a few lines later when `dialog.pack()` is called because `dialog` is `null`.  This issue is still present today.

The latest code uses the new `swinglib.JDialogBuilder` class, which enforces that the parent frame is non-`null` via a precondition check.  I'm assuming that's why the `parent != null` check was left intact.  However, the NPE described in the previous paragraph will still occur.  If an NPE (i.e. a precondition failure) is the desired behavior, then it might be cleaner to simply remove the `parent != null` check and let `JDialogBuilder#parentFrame()` throw the exception.

It's not clear if there is a code path that will call this method with a `null` parent frame.  The Javadocs for `JDialog(Frame, String, boolean)` seem to imply that one cannot specify a `null` parent frame.  Presumably, this is why the original author went to the trouble to add the `parent != null` check.  However, the source code for Java 8 suggests that `null` is handled by using a prefab `Frame` from the `SwingUtilities#getSharedOwnerFrame()` method.  It's possible the non-`null` parent frame requirement was relaxed in a later Java release after Swing was first introduced?

We may have to re-evaluate the precondition requirement on `JDialogBuilder#parentFrame()`.  It may be that it is an over-constraint.  @DanVanAtta, what do you think?

Regardless of whether or not `JDialogBuilder` should require a non-`null` parent frame, the following patch can be used to fix this issue:

```diff
diff --git a/src/main/java/games/strategy/triplea/ui/ProductionPanel.java b/src/main/java/games/strategy/triplea/ui/ProductionPanel.java
index 3ee2901..d76afb6 100644
--- a/src/main/java/games/strategy/triplea/ui/ProductionPanel.java
+++ b/src/main/java/games/strategy/triplea/ui/ProductionPanel.java
@@ -82,20 +82,11 @@ public class ProductionPanel extends JPanel {
    */
   public IntegerMap<ProductionRule> show(final PlayerID id, final JFrame parent, final GameData data, final boolean bid,
       final IntegerMap<ProductionRule> initialPurchase) {
-    if (parent != null) {
-      dialog = JDialogBuilder.builder()
-          .parentFrame(parent)
-          .contents(this)
-          .title("Produce")
-          .build();
-      dialog.pack();
-      dialog.setLocationRelativeTo(parent);
-      dialog.setVisible(true);
-      // making the dialog visible will block until it is closed
-      dialog.dispose();
-
-
-    }
+    dialog = JDialogBuilder.builder()
+        .parentFrame(parent)
+        .contents(this)
+        .title("Produce")
+        .build();
     this.bid = bid;
     this.data = data;
     this.initRules(id, data, initialPurchase);
```

In the case where `JDialogBuilder` continues to require a non-`null` parent frame, this will fail fast and throw an NPE (and probably requires updating the Javadocs for `ProductionPanel#show()` to explicitly indicate it requires a non-`null` parent frame).  If we change `JDialogBuilder` to allow a `null` parent frame, the code will end up using the Swing shared owner frame, and no NPE will be thrown.

**Comment by ssoloff** [comment]
Just verifying that passing a `null` parent frame to the `JDialog(Frame, String, boolean)` constructor works just fine.  The only difference in behavior from the non-`null` case is that the dialog gets its own item in the system tray and uses the default Java icon (similar to what we've observed when creating other TripleA windows with a `null` reference instead of the main frame window).

**Comment by ron-murhammer** [comment]
@ssoloff Yeah, this is pretty bad. Little hard to follow all the changes from the previous PR here to understand why this is broken. I'll let @DanVanAtta take a look since he probably has more context here.

**Comment by ssoloff** [comment]
@DanVanAtta Bumping to see if you have an opinion on whether `JDialogBuilder` should or should not allow a `null` parent frame based on the information in the issue description.

**Comment by DanVanAtta** [comment]
@ssoloff responded in #2209, in general though, defaults should be reasonable and preferred. Given that, if we do specify a parent frame as a builder param, then it should be non-null. Clients should know when they are handling null (nonsense data).

**Comment by DanVanAtta** [comment]
To clarify:
IMO this is okay:
```
    dialog = JDialogBuilder.builder()
.build()
```
But not this:
```
    dialog = JDialogBuilder.builder()
.parentFrame(null) // << this is bad, forces client to make a useless api call
.build()
```

**Comment by ssoloff** [comment]
Thanks for the feedback, @DanVanAtta.  I completely agree with your suggestion and will update #2209 accordingly.
